packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -1742,17 +1764,19 @@ function createChildReconciler(
     // not as a fragment. Nested arrays on the other hand will be treated as
     // fragment nodes. Recursion happens at the normal flow.
 
-    // Handle top level unkeyed fragments as if they were arrays.
-    // This leads to an ambiguity between <>{[...]}</> and <>...</>.
+    // Handle top level unkeyed fragments without refs (enableFragmentRefs)
+    // as if they were arrays. This leads to an ambiguity between <>{[...]}</> and <>...</>.
     // We treat the ambiguous cases above the same.
     // We don't use recursion here because a fragment inside a fragment
     // is no longer considered "top level" for these purposes.
-    const isUnkeyedTopLevelFragment =
+    const isUnkeyedUnrefedTopLevelFragment =
       typeof newChild === 'object' &&
       newChild !== null &&
       newChild.type === REACT_FRAGMENT_TYPE &&
-      newChild.key === null;
-    if (isUnkeyedTopLevelFragment) {
+      newChild.key === null &&
+      (enableFragmentRefs ? newChild.props.ref === undefined : true);
```

Comment:
This is subtle but this means that `null` and `undefined` are different and will reconcile differently. Which is technically a breaking change.

It also means that `<div ref={maybeUndefined && ref}>{children}</div>` will clear the state of the whole subtree when it switches.

If it was `ref === null || ref === undefined` then it would clear state even if you did `<div ref={c ? ref : null}>` which also doesn't seem right.

So I think what you have is probably the best compromise.

Reply:
This is specific to Fragment components, right? So your example of clearing state of the `<div />` subtree would be limited to `<Fragment ref={}>` subtree?

I basically want to know here whether a ref prop was declared at all. But I don't think we can distinguish a `maybeUndefined` ref value.